app_dart/lib/src/request_handlers/get_benchmarks.dart

```diff
@@ -30,26 +33,102 @@ class GetBenchmarks extends RequestHandler<Body> {
 
   @override
   Future<Body> get() async {
-    final String branch = request.uri.queryParameters[branchParam] ?? 'master';
-    const int maxRecords = 50;
+    const String master = 'master';
+
+    final int maxRecords = config.maxRecords;
+    final String branch = request.uri.queryParameters[branchParam] ?? master;
     final DatastoreService datastore = datastoreProvider(config.db);
-    final DatastoreDB db = datastore.db;
+    final List<Map<String, dynamic>> benchmarks = <Map<String, dynamic>>[];
+
+    Map<String, Result> releaseBranchMap = <String, Result>{};
+    Map<String, Result> masterMap = <String, Result>{};
+    int masterLimit = maxRecords;
+    int timestamp = DateTime.now().millisecondsSinceEpoch;
+
+    /// Query all commits of the release branch first. Then calcalute the
+    /// number of commits to retrieve from master branch, and obtain
+    /// the starting [timestamp] to filter master commits.
+    if (branch != master) {
+      final List<Commit> releaseBranchCommits = await datastore
+          .queryRecentCommits(limit: maxRecords, branch: branch)
+          .toList();
+      releaseBranchMap = await _getBenchmarks(releaseBranchCommits.length,
+          branch, datastore, releaseBranchCommits, benchmarks, timestamp);
+      masterLimit = maxRecords > releaseBranchCommits.length
+          ? maxRecords - releaseBranchCommits.length
+          : 0;
+      timestamp = releaseBranchCommits.last.timestamp;
+    }
+
+    /// Query all remaining commits from master.
+    if (branch == master || masterLimit > 0) {
+      /// `+1` is to guarantee picking up the master commit, from which
+      /// the release branch is derived.
+      final List<Commit> masterCommits = await datastore
+          .queryRecentCommits(timestamp: timestamp + 1, limit: masterLimit)
+          .toList();
```

Comment:
What +1 means in this context? 1 ms, 1m? are we introducing a race condition?

Reply:
timestamp is in milliseconds, which should allow only the commit with same timestamp as the release branch.

Reply:
We are mixing timestamp and milliseconds since epoch. We should use one or the other and name properties parameters accordingly.

Reply:
It has been designed as to use `timestamp` variable name in cocoon to hold values of milliseconds since epoch for all APIs, and in datastore as well. If we want to distinguish them, we may need a separate PR to change from the base.

Reply:
My comment was not about changing the current types but to make the variable names consistent with the content of them. In this case name your variable something like currentMsSinceEpoch instead of timestamp to make it consistent with the content.